src/System.IO.Pipelines.Networking.Tls/SecureConnectionContext.cs

```diff
@@ -0,0 +1,343 @@
+        {
+            get { return _trailerSize; }
+            set { _trailerSize = value; }
+        }
+
+        public SSPIHandle ContextHandle => _contextPointer;
+        public bool IsServer => _securityContext.IsServer;
+        public CipherInfo CipherInfo => Interop.GetCipherInfo(_contextPointer);
+
+        /// <summary>
+        /// Without a payload from the client the server will just return straight away.
+        /// </summary>
+        /// <param name="writeBuffer"></param>
+        public Task ProcessContextMessageAsync(IPipelineWriter writeBuffer)
+        {
+            return ProcessContextMessageAsync(default(ReadableBuffer), writeBuffer);
+        }
+
+        /// <summary>
+        /// Processes the tokens or cipher change messages from a client and can then return server messages for the client
+        /// </summary>
+        /// <param name="readBuffer"></param>
+        /// <param name="writePipeline"></param>
+        public Task ProcessContextMessageAsync(ReadableBuffer readBuffer, IPipelineWriter writePipeline)
+        {
+            var handleForAllocation = default(GCHandle);
+            try
+            {
+                var output = new SecurityBufferDescriptor(2);
+                var outputBuff = stackalloc SecurityBuffer[2];
+                outputBuff[0] = new SecurityBuffer(null, 0, SecurityBufferType.Token);
+                outputBuff[1] = new SecurityBuffer(null, 0, SecurityBufferType.Alert);
+                output.UnmanagedPointer = outputBuff;
+
+                var handle = _securityContext.CredentialsHandle;
+                SSPIHandle localhandle = _contextPointer;
+                void* contextptr;
+                void* newContextptr;
+                if (_contextPointer.handleHi == IntPtr.Zero && _contextPointer.handleLo == IntPtr.Zero)
+                {
+                    contextptr = null;
+                    newContextptr = &localhandle;
+                }
+                else
+                {
+                    contextptr = &localhandle;
+                    newContextptr = null;
+                }
+                var unusedAttributes = default(ContextFlags);
+                SecurityBufferDescriptor* pointerToDescriptor = null;
+
+                if (readBuffer.Length > 0)
+                {
+                    var input = new SecurityBufferDescriptor(2);
+                    var inputBuff = stackalloc SecurityBuffer[2];
+                    inputBuff[0].size = readBuffer.Length;
+                    inputBuff[0].type = SecurityBufferType.Token;
+
+                    if (readBuffer.IsSingleSpan)
+                    {
+                        void* arrayPointer;
+                        readBuffer.First.TryGetPointer(out arrayPointer);
+                        inputBuff[0].tokenPointer = arrayPointer;
+                    }
+                    else
+                    {
+                        if (readBuffer.Length <= SecurityContext.MaxStackAllocSize)
```

Comment:
maybe, maybe not, this has a finite depth of methods it will call. We have all the bytes for a frame at this point, and we know the methods we will call (basically straight into OpenSsl or Sspi and then unwind). There could be a debate about the safe size, is there a way to tell how much stack is left?